core/src/main/java/org/elasticsearch/common/Randomness.java

```diff
@@ -0,0 +1,148 @@
+
+    static {
+        Method maybeCurrentMethod;
+        Method maybeGetRandomMethod;
+        try {
+            Class<?> clazz = Class.forName("com.carrotsearch.randomizedtesting.RandomizedContext");
+            maybeCurrentMethod = clazz.getMethod("current");
+            maybeGetRandomMethod = clazz.getMethod("getRandom");
+        } catch (Throwable t) {
+            maybeCurrentMethod = null;
+            maybeGetRandomMethod = null;
+        }
+        currentMethod = maybeCurrentMethod;
+        getRandomMethod = maybeGetRandomMethod;
+    }
+
+    private Randomness() {}
+
+    /**
+     * Provides a reproducible source of randomness seeded by a long
+     * seed in the settings with the key setting.
+     *
+     * @param settings the settings containing the seed
+     * @param setting  the key to access the seed
+     * @return a reproducible source of randomness
+     */
+    public static Random get(Settings settings, String setting) {
+        Long maybeSeed = settings.getAsLong(setting, null);
+        if (maybeSeed != null) {
+            return new Random(maybeSeed);
+        } else {
+            return get();
+        }
+    }
+
+    /**
+     * Provides a source of randomness that is reproducible when
+     * running under the Elasticsearch test suite, and otherwise
+     * produces a non-reproducible source of randomness. Reproducible
+     * sources of randomness are created when the system property
+     * "tests.seed" is set, is parseable as a hexadecimal long, and the
+     * security policy allows reading this system property. Otherwise,
+     * non-reproducible sources of randomness are created.
+     *
+     * @return a source of randomness
+     * @throws IllegalArgumentException if "tests.seed" is set but the
+     *                                  system property with that key
+     *                                  is not a string of hexadecimal
+     *                                  digits.
+     */
+    public static Random get() {
+        if (currentMethod != null && getRandomMethod != null) {
+            if (getTestsSeedProperty() != null) {
+                try {
+                    Object randomizedContext = currentMethod.invoke(null);
+                    return (Random) getRandomMethod.invoke(randomizedContext);
+                } catch (IllegalAccessException | InvocationTargetException e) {
+                    // unexpected, bail
+                    throw new IllegalStateException("running tests but failed to invoke RandomizedContext#getRandom", e);
+                }
+            } else {
+                throw new IllegalStateException("found test classes on the classpath but tests.seed is not set");
+            }
+        }  else {
+            return getWithoutSeed();
+        }
+    }
+
+    private static String getTestsSeedProperty() {
+        String property = null;
+        try {
+            property = (String) BootstrapInfo.getSystemProperties().get("tests.seed");
```

Comment:
Why not just use System.getProperty("tests.seed") ? is this an intellij auto-complete thing? I see this anti-pattern quite often and i wonder why its done.